core/src/main/scala/kafka/admin/AdminUtils.scala

```diff
@@ -300,14 +307,14 @@ object AdminUtils extends Logging with AdminUtilities {
     // check if manual assignment has the right replication factor
     val unmatchedRepFactorList = newPartitionReplicaList.values.filter(p => p.size != existingReplicaListForPartitionZero.size)
     if (unmatchedRepFactorList.nonEmpty)
-      throw new AdminOperationException("The replication factor in manual replication assignment " +
-        " is not equal to the existing replication factor for the topic " + existingReplicaListForPartitionZero.size)
-
-    info("Add partition list for %s is %s".format(topic, newPartitionReplicaList))
-    val partitionReplicaList = existingPartitionsReplicaList.map(p => p._1.partition -> p._2)
-    // add the new list
-    partitionReplicaList ++= newPartitionReplicaList
-    AdminUtils.createOrUpdateTopicPartitionAssignmentPathInZK(zkUtils, topic, partitionReplicaList, update = true)
+      throw new InvalidReplicaAssignmentException(s"Existing topic replication factor of ${existingReplicaListForPartitionZero.size}, but manual replication assignment would imply replication factor(s) of ${unmatchedRepFactorList.map(_.size).mkString(",")}.")
+    if (!validateOnly) {
+      info(s"Add partition list for $topic is $newPartitionReplicaList.")
+      val partitionReplicaList = existingPartitionsReplicaList.map(p => p._1.partition -> p._2)
```

Comment:
Maybe we can take the chance and clean-up this code a bit. In general `_1` and `_2` are not very readable. The following is clearer:

```scala
existingPartitionsReplicaList.map { case (tp, replicas) => (tp.partition, replicas) }
```